jupyter_collaboration/rooms.py

```diff
@@ -89,64 +103,59 @@ async def initialize(self) -> None:
+            read_from_source = True
+
+        # if YStore updates and source file are out-of-sync, resync updates with source
+        if read_from_source:
+            self._emit(LogLevel.INFO, "load", "Content loaded from disk.")
+            self.log.info(
+                "Content in room %s loaded from file %s", self._room_id, self._file.path
+            )
+            self._document.source = model["content"]
+
+            if self.ystore:
+                await self.ystore.encode_state_as_update(self.ydoc)
 
-            async with self._update_lock:
-                # try to apply Y updates from the YStore for this document
-                read_from_source = True
-                if self.ystore is not None:
-                    try:
-                        await self.ystore.apply_updates(self.ydoc)
-                        self._emit(
-                            LogLevel.INFO,
-                            "load",
-                            "Content loaded from the store {}".format(
-                                self.ystore.__class__.__qualname__
-                            ),
-                        )
-                        self.log.info(
-                            "Content in room %s loaded from the ystore %s",
-                            self._room_id,
-                            self.ystore.__class__.__name__,
-                        )
-                        read_from_source = False
-                    except YDocNotFound:
-                        # YDoc not found in the YStore, create the document from the source file (no change history)
-                        pass
-
-                if not read_from_source:
-                    # if YStore updates and source file are out-of-sync, resync updates with source
-                    if self._document.source != model["content"]:
-                        # TODO: Delete document from the store.
-                        self._emit(
-                            LogLevel.INFO, "initialize", "The file is out-of-sync with the ystore."
-                        )
-                        self.log.info(
-                            "Content in file %s is out-of-sync with the ystore %s",
-                            self._file.path,
-                            self.ystore.__class__.__name__,
-                        )
-                        read_from_source = True
-
-                if read_from_source:
-                    self._emit(LogLevel.INFO, "load", "Content loaded from disk.")
-                    self.log.info(
-                        "Content in room %s loaded from file %s", self._room_id, self._file.path
-                    )
-                    self._document.source = model["content"]
-
-                    if self.ystore:
-                        await self.ystore.encode_state_as_update(self.ydoc)
-
-                self._document.dirty = False
-                self.ready = True
-                self._emit(LogLevel.INFO, "initialize", "Room initialized")
```

Comment:
Aside from removing the locks and the `self._document.dirty = False` statements, there are no code changes in this section. Most of my substantive code changes are in the `_maybe_save_document()` method.